### PR TITLE
feat: introduce basic ssl connection

### PR DESCRIPTION
**Pull request by BowenXiao1999**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
Security and SSL is a big topic and may involves a lot tasks. This PR just finishes the very basic function.

For test
`./risedev p`
`psql -p 4566 -d dev -h localhost -U root --set=sslmode=verify-full`
You can see that 
`SSL connection (protocol: TLSv1.3, cipher: TLS_AES_256_GCM_SHA384, bits: 256, compression: off)`

For code
Build up needed Ssl context when protocol set up, and turn this context into ssl stream when received ssl request. 
Add a Conn enum to wrap up Unencrypted and encrypted. Simply static dispatch the method call.
Self-signed CA cert and key for tests under test folder for e2e test.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
ref #5604

### CODE REVIEW

File: src/utils/pgwire/src/pg_protocol.rs [hunk 1]
```diff
@@ -57,6 +61,29 @@ where
...
+}
+
+impl TlsConfig {
+    fn new_default() -> Self {
+        Self {
+            // Now the demo crt and key are hard code generated via simple self-signed CA.
+            // In future it should change to configure by user.
+            cert: "src/utils/pgwire/tests/ssl/demo.crt".into(),
+            key: "src/utils/pgwire/tests/ssl/demo.key".into(),
+        }
```
Comment (neverchanje) [inline src/utils/pgwire/src/pg_protocol.rs:85]: What will the demo key and cert be used for?

**Review by wangrunji0408 — APPROVED** [review]
LGTM

File: src/utils/pgwire/src/pg_protocol.rs (hunk 1 above)
Comment (BowenXiao1999) [reply src/utils/pgwire/src/pg_protocol.rs:85]: Used for SSL handshake. There must be certificate and private key.

**Review by mikechesterwang — APPROVED** [review]
LGTM

**Comment by skyzh** [comment]
```


thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: ErrorStack([Error { code: 33558530, library: "system library", function: "fopen", reason: "No such file or directory", file: "crypto/bio/bss_file.c", line: 288, data: "fopen('src/utils/pgwire/tests/ssl/demo.key','r')" }, Error { code: 537346050, library: "BIO routines", function: "file_ctrl", reason: "system lib", file: "crypto/bio/bss_file.c", line: 290 }, Error { code: 336265218, library: "SSL routines", function: "SSL_CTX_use_PrivateKey_file", reason: "system lib", file: "ssl/ssl_rsa.c", line: 521 }])', src/utils/pgwire/src/pg_protocol.rs:767:10
--
  | note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  | note: run with `MADSIM_TEST_SEED=1665763720914771120` environment variable to reproduce this error
  | and make sure `MADSIM_CONFIG_HASH=1AD2E4ABC41F3553`
 

<br class="Apple-interchange-newline" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">
```

Probably need static link for building RisingWave in order not to open dynamic libraries?

**Comment by skyzh** [comment]
Need to ignore this case from deterministic simulation, or support open files. cc @wangrunji0408 Thanks for looking into this.

**Comment by wangrunji0408** [comment]
I found this error only occurs in deterministic scaling test, the other two deterministic test tasks are passed. So could the reason be that the scaling test runs on a different working directory? cc @BugenZhao

**Comment by skyzh** [comment]
Sounds great!

**Comment by BugenZhao** [comment]
> So could the reason be that the scaling test runs on a different working directory?

Oops. It's possible. I archive the unit tests with `nextest` in the build phase and I'm not sure where they will be extracted and run. 🤔

**Comment by BugenZhao** [comment]
As the key and cert are hard-coded now and not production ready, can we encode them in the program and use `tempfile` in runtime? So we don't need to worry about the working directory. For example, user may find RisingWave can't work if they want to manually run it in a different working directory.

**Comment by BowenXiao1999** [comment]
Actually I can just turn the `do not read file` into an error so that non-e2e-test will not have ssl connection. After investigating, I think this is the most convenient way to bypass this problem.

**Comment by BowenXiao1999** [comment]
> As the key and cert are hard-coded now and not production ready, can we encode them in the program and use tempfile in runtime? So we don't need to worry about the working directory. For example, user may find RisingWave can't work if they want to manually run it in a different working directory.

Thanks for sharing. I will investigate this.

**Comment by BowenXiao1999** [comment]
The root cause is cargo run/cargo test will start from the closet Cargo.toml root, while the hard code start from project root. So I think moving tests folder can only temporailly solve this, we need to find a better way.

**Pull request merged**